Wire-Up CI - Fix a few nagging issues

msheiny opened the pull request:
This PR hooks up the repo to circleci for automated testing and uses a strategy very
similar to some of our on-going web repos for testing. Right now it is configured to test against 
https://github.com/littleweaver/littleweaverweb.com for two reasons:

* its an open repo (no deploy keys needed)
* uses nodejs 6.x (our only open www repo, securethenews, uses 4.x )

In order to test locally just run `make`, and then `make tests`. Hoorayyyy.

msheiny committed:
optional - utilize rsync instead of git 4 code

Basically, instead of pulling down the django code from a git repo
this will utilize rsync to push a defined local directory there
instead. Really useful in certain situations under CI and local testing
where the git repo is private but you dont want to go through the deploy_key
dance.

msheiny committed:
Allow override of npm install command

In case any particular flags or options need to get passed along.

msheiny committed:
Allow over-ride of rsync options

Allows customization of rsync options via variable override.

msheiny committed:
Update git pull detection change logic

This was really wonky before, cleaned up the logic a bit so it should be more
reliable.

Why did I also add in those `become` lines? two words `ansible-lint`

msheiny committed:
Allow override of django settings file

This allows a playbook to specify a local settings file that would ideally
over-ride settings files in the repo. Typically this plays nice when the repo
you are deploying has an ignore exception for a settings file. This is usually
best practice on a django repo to ignore a local.py settings file for developers
to hack on.

msheiny committed:
Dockerize testing of playbook, hook-up CI

Thie commit rips out the previous molecule reliance on vagrant, re-organizes
the repo, swaps in docker pieces, and now uses a live site (littleweaverweb.com)
repository to test against.

msheiny committed:
Update readme to reflect CI/Testing changes

Lots of changes have occurred with testing and deployment options since this
readme was originally fleshed out. Lets update it to make it a little
friendlier.

msheiny committed:
CircleCI fixes - temp swap-out docker CI image

Trying to shake out issues building `cryptography` with the latest image up on
quay.io - falling back to the `msheiny` one on dockerhub seems to do the trick
for the time being.

msheiny commented:
One note is that the littleweaver repo this is wired up against doesnt use a `package.json` to track node dependencies so we arent really testing an npm install from the repo directly. This has been a problematic piece of functionality (had to use `raw` versus the ansible module). So ... thats just one thing to note going forward. If we open-source another one of our websites in the future probably want to swap this to test against that instead.

conorsch reviewed (approved):
Worked like a charm during local testing. Very nice! :tophat:

msheiny committed:
Resolve weird set_fact recursive variable issue

Sooo I'm actually not 100% this is going to fix it, but I saw a weird issue in
prod related to this while trying to run a recent deploy. Crossing fingers here.

msheiny committed:
Idempotency fix for node dir permissions

npm is weird and I dont see this as a useful change indicator if the ansible
playbook is always going to tell me that the npm folder task changed. I'm
masking over the change status from ansible going forward. Dont think its worth
investigating this nit.

msheiny commented:
I'm going to merge as-is and open a ticket for the new problem plaguing prod ;) Its unrelated to this change and a really weird one. Yayyyyyyyy. I only mentioned that here cause I tried to push commits to fix it but now I think i know what it is.
